tests/unit/test_dimension_utilities.py

```diff
@@ -304,6 +320,165 @@ def test_prefetch_dimension_variables(self, mock_get_opendap_nc4):
+        """
+        varinfo_bounds = VarInfoFromDmr(
+            'tests/data/ATL16_prefetch_bnds.dmr'
+        )
+
+        with self.subTest('Variable has cell alignment and bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_edge_has_bnds')))
+
+        with self.subTest('Variable has no cell alignment and has bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_no_edge_has_bnds')))
+
+        with self.subTest('Variable has cell alignment and no bounds'):
+            self.assertTrue(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_edge_no_bnds')))
+
+        with self.subTest('Variable has no cell alignment and no bounds'):
+            self.assertFalse(needs_bounds(varinfo_bounds.get_variable(
+                '/variable_no_edge_no_bnds')))
+
+    def test_get_bounds_array(self):
+        """ Ensure that the expected bounds array is created given
+            the input dimension variable values.
+
+        """
+        prefetch_dataset = Dataset('tests/data/ATL16_prefetch.nc4', 'r')
+        dimension_path = '/npolar_grid_lat'
+
+        expected_bounds_array = self.bounds_array
+
+        assert_array_equal(get_bounds_array(prefetch_dataset,
+                                            dimension_path),
+                           expected_bounds_array)
+
+    def test_write_bounds(self):
+        """ Ensure that bounds data array is written to the dimension
+            dataset, both when the dimension variable is in the root group
+            and in a nested group.
+
+        """
+        varinfo_prefetch = VarInfoFromDmr('tests/data/ATL16_prefetch_group.dmr')
+        prefetch_dataset = Dataset('tests/data/ATL16_prefetch_group.nc4', 'r+')
+
+        # Expected variable contents in file.
+        expected_bounds_data = self.bounds_array
+
+        with self.subTest('Dimension variable is in the root group'):
+            root_variable_full_path = '/npolar_grid_lat'
+            root_varinfo_variable = varinfo_prefetch.get_variable(
+                root_variable_full_path)
+            root_variable_name = 'npolar_grid_lat'
+            root_bounds_name = root_variable_name + '_bnds'
+
+            write_bounds(prefetch_dataset, root_varinfo_variable)
+
+            resulting_bounds_root_data = prefetch_dataset.variables[
+                root_bounds_name][:]
+            assert_array_equal(resulting_bounds_root_data,
+                               expected_bounds_data)
+            # Check that bounds variable exists in the root group.
+            self.assertTrue(prefetch_dataset.variables[root_variable_name])
+            # Check that bounds variable has expected attributes.
+            # (This is only the 'bounds' attribute, since other attributes
+            # are not required.)
+            root_bounds_varinfo_variable = varinfo_prefetch.get_variable(
+                root_variable_full_path)
```

Comment:
Disclaimer, I might be getting mixed up here:

```
# L423
 root_varinfo_variable = varinfo_prefetch.get_variable(
     root_variable_full_path)
```

and

```
# L439
root_bounds_varinfo_variable = varinfo_prefetch.get_variable(
    root_variable_full_path)
```

So aren't the next two assertions doing the same test?

Also `root_bounds_varinfo_variable` kind of implies that you are getting the `bounds` variable (e.g., `/npolar_grid_lat_bnds`) rather than the dimension variable.

Reply:
These variable names are... very difficult. I can maybe try to reorganize them to make them make at least a little bit more sense.

Yea, those are doing the same thing and I did mean to have a `root_bounds_varinfo_variable`. Wasn't this test for checking out the `bounds` variable? I must be turned around. Looking below, I guess I thought #3 was what this test was doing.